File: EmojiPanel/src/main/java/com/binlee/emoji/HttpAdapter.java
package com.binlee.emoji;

import com.binlee.emoji.compat.HttpEngine;
import com.binlee.emoji.third.OkHttpEngine;

public class HttpAdapter {

    private static HttpEngine sEngine;

    private HttpAdapter() {
        throw new AssertionError("no instance.");
    }

    public static void setHttpEngine(HttpEngine engine) {
        if (engine == null) {
            throw new NullPointerException("engine is null");
        }
        sEngine = engine;
    }

    public static HttpEngine engine() {
        if (sEngine == null) {
            sEngine = new OkHttpEngine();
            // sEngine = new DefaultHttpEngine();
            // sEngine = new SocketHttpEngine();
        }
        return sEngine;
    }
}

File: EmojiPanel/src/main/java/com/binlee/emoji/HttpAdapter.kt
package com.binlee.emoji

import com.binlee.emoji.compat.DefaultHttpEngine
import com.binlee.emoji.compat.HttpEngine
import com.binlee.emoji.third.OkHttpEngine
import com.binlee.emoji.third.SocketHttpEngine

class HttpAdapter private constructor() {

    companion object {

        private var sEngine: HttpEngine<*, *>? = null

        fun setHttpEngine(engine: HttpEngine<*, *>?) {
            if (engine == null) {
                throw NullPointerException("engine is null")
            }
            sEngine = engine
        }

        fun engine(): HttpEngine<*, *> {
            if (sEngine == null) {
                sEngine = OkHttpEngine()
                // sEngine = DefaultHttpEngine()
                // sEngine = SocketHttpEngine()
            }
            return sEngine!!
        }
    }

    init {
        throw AssertionError("no instance.")
    }
}